File: cpp/linked-lists/0160-intersection-of-two-linked-lists.cpp
#include <structures.h>


ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
    ListNode *curA = headA;
    ListNode *curB = headB;
    int lenA = 0;
    int lenB = 0;
    
    while (curA->next) {
        lenA++;
        curA = curA->next;
    }
    
    while (curB->next) {
        lenB++;
        curB = curB->next;
    }
    
    curA = headA;
    curB = headB;

    if (lenA > lenB) {
        for (int i = 0; i < lenA - lenB; i++) curA = curA->next; 
    }

    if (lenA < lenB) {
        for (int i = 0; i < lenB - lenA; i++) curB = curB->next;
    }
    
    while (curA && curB) {
        if (curA == curB) return curA;;
        curA = curA->next;
        curB = curB->next;
    }
    
    return nullptr;
}

File: js/linked-lists/0160-intersection-of-two-linked-lists.js
var getIntersectionNode = function(headA, headB) {
    var lenA = 0;
    var curA = headA;
    while (curA) {
        curA = curA.next;
        lenA++;
    }
    
    var lenB = 0;
    var curB = headB;
    while (curB) {
        curB = curB.next;
        lenB++;
    }
    
    curA = headA;
    curB = headB
    
    if (lenA > lenB) {
        for (var i = 0; i < lenA - lenB; i++) {
            curA = curA.next;
        }    
    } else if (lenB > lenA) {
        for (var i = 0; i < lenB - lenA; i++) {
            curB = curB.next;
        }
    }
    
    while (curA && curB) {
        if (curA === curB) {
            return curA;
        }
        curA = curA.next;
        curB = curB.next;
    }
    
    return null;
};